File: CS/E4787/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using DevExpress.Web.Mvc;
using E4787.Models;

namespace E4787.Controllers {
    public class HomeController : Controller {
        public ActionResult Index() {
            return View();
        }

        [ValidateInput(false)]
        public ActionResult GridViewPartial() {
            var model = GetData();
            return PartialView("_GridViewPartial", model);
        }

        public ActionResult AddNew(Entry e) {
            ViewData["newKey"] = e.ID;

            var model = GetData().ToList();
            model.Add(e);
            return PartialView("_GridViewPartial", model);
        }

        private IEnumerable<Entry> GetData() {
            return Enumerable.Range(0, 5).Select(i => new Entry{ ID=i, Text = "Text " + i });
        }

    }
}

File: VB/E4787/Controllers/HomeController.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.Mvc
Imports DevExpress.Web.Mvc
Imports E4787.Models

Namespace E4787.Controllers
	Public Class HomeController
		Inherits Controller

		Public Function Index() As ActionResult
			Return View()
		End Function

		<ValidateInput(False)>
		Public Function GridViewPartial() As ActionResult
			Dim model = GetData()
			Return PartialView("_GridViewPartial", model)
		End Function

		Public Function AddNew(ByVal e As Entry) As ActionResult
			ViewData("newKey") = e.ID

			Dim model = GetData().ToList()
			model.Add(e)
			Return PartialView("_GridViewPartial", model)
		End Function

		Private Function GetData() As IEnumerable(Of Entry)
			Return Enumerable.Range(0, 5).Select(Function(i) New Entry With {
				.ID=i,
				.Text = "Text " & i
			})
		End Function

	End Class
End Namespace